### PR TITLE
Move web tier to use autoscaling groups

### PR DESCRIPTION
**Pull request by mpeamma**

### CODE REVIEW

File: webtier.tf [hunk 1]
```diff
@@ -57,3 +30,86 @@ resource "aws_lb_listener" "3tier-http-listener" {
...
+  name_prefix   = "3tier-web-asg-template"
+  image_id      = "${var.web_tier_ami}"
+  instance_type = "t2.micro"
+  vpc_security_group_ids = ["${aws_security_group.3tier-web-sg.id}"]
+  user_data = "${base64encode(var.web_tier_userdata)}"
+  key_name  = "${var.keypair_name}"
+}
+
+resource "aws_autoscaling_group" "3tier-web-asg" {
+  vpc_zone_identifier = ["${aws_subnet.3tier-private-subnets.*.id[0]}","${aws_subnet.3tier-private-subnets.*.id[1]}","${aws_subnet.3tier-private-subnets.*.id[2]}",]
```
Comment (sean-nixon) [inline webtier.tf]: We can't hard code each element like this. Eventually the count will be a variable

File: webtier.tf [hunk 2]
```diff
@@ -57,3 +30,86 @@ resource "aws_lb_listener" "3tier-http-listener" {
...
+  protocol          = "all"
+  cidr_blocks       = ["0.0.0.0/0"]
+}
+
+resource "aws_launch_template" "3tier-web-asg-template" {
+  name_prefix   = "3tier-web-asg-template"
+  image_id      = "${var.web_tier_ami}"
+  instance_type = "t2.micro"
+  vpc_security_group_ids = ["${aws_security_group.3tier-web-sg.id}"]
+  user_data = "${base64encode(var.web_tier_userdata)}"
```
Comment (sean-nixon) [inline webtier.tf]: Remove the base64 encoding. This variable does the encoding automatically.

File: webtier.tf [hunk 3]
```diff
@@ -57,3 +30,86 @@ resource "aws_lb_listener" "3tier-http-listener" {
...
+  key_name  = "${var.keypair_name}"
+}
+
+resource "aws_autoscaling_group" "3tier-web-asg" {
```
Comment (sean-nixon) [inline webtier.tf]: Let's add some basic health check configuration here so we can make them variables later on. Keeping whatever the defaults are is probably fine. Just want to make it explicit so we know what we can parameterize.

File: webtier.tf (hunk 2 above)
Comment (mpeamma) [reply webtier.tf]: I was getting errors through terraform about the base64 encoding without this

File: webtier.tf (hunk 1 above)
Comment (mpeamma) [reply webtier.tf]: Does terraform support syntax that would just say "get all of the subnets in this list"?

File: webtier.tf (hunk 2 above)
Comment (sean-nixon) [reply webtier.tf]: Odd. There's a user_data_base64 parameter as well that's intended for already base64 encoded data. I'm curious is the user data getting set as expected with this configuration?

File: webtier.tf (hunk 1 above)
Comment (sean-nixon) [reply webtier.tf]: I'm wondering if `"${aws_subnet.3tier-private-subnets.*.id}" would work. Terraform 0.11 is weird how it deals with this though so I don't know.

Comment (sean-nixon) [reply webtier.tf]: Yeah, I think the above will work. You may need to remove the `[]`

File: webtier.tf (hunk 2 above)
Comment (mpeamma) [reply webtier.tf]: It looks like they are doing this in the terraform docs https://www.terraform.io/docs/providers/aws/r/launch_template.html

Comment (sean-nixon) [reply webtier.tf]: Ah, I was looking at the launch configuration template. It looks like it is expecting an already base64 encoded string. Looks good then.

File: webtier.tf (hunk 3 above)
Comment (mpeamma) [reply webtier.tf]: I added the health check explicitly to the target group

File: webtier.tf (hunk 1 above)
Comment (mpeamma) [reply webtier.tf]: I ended up needing to keep the `[]` but that worked

File: webtier.tf (hunk 3 above)
Comment (sean-nixon) [reply webtier.tf]: Pretty sure auto-scaling groups have separate health check configuration. I'll look into it

**Pull request merged**